# 01. Data Understanding - Automobile Loan Default Prediction
**Module:** IT3051 – Fundamentals of Data Mining  
**Objective:** Gain a comprehensive initial understanding of the Automobile Loan Default raw dataset.

---

### What is "Data Understanding"?
In data mining (CRISP-DM methodology), **Data Understanding** is the foundational phase where we inspect raw data to:
1. Understand the size and structure of the dataset.
2. Check the data types of individual features.
3. Identify missing or corrupted data.
4. Detect data quality anomalies (e.g., numbers stored as text strings).

This notebook explores the raw loan application data (`Train_Dataset.csv`) step-by-step with beginner-friendly explanations for undergraduate study.


## 1. Importing Libraries
We begin by importing **Pandas** and **NumPy**.
- `pandas` (`pd`): The standard Python library for data manipulation and analysis using tables called **DataFrames**.
- `numpy` (`np`): A numerical computing library providing support for mathematical operations and `NaN` (Not a Number) representations.
- `os`: Allows us to check file paths safely regardless of where Jupyter is launched.


In [9]:
import os
import pandas as pd
import numpy as np

# Set display options so we can inspect all columns comfortably
pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 1000)

print("Libraries imported successfully!")


Libraries imported successfully!


## 2. Loading the Dataset
We load the raw CSV file using `pd.read_csv()`.

> **Why `low_memory=False`?**  
> Pandas reads large CSV files in chunks and tries to guess the data type of each column based on early rows. If a column contains mostly numbers in early rows but has symbols (such as `$` or `x`) in later rows, Pandas will issue a `DtypeWarning`. Specifying `low_memory=False` instructs Pandas to load the entire column into memory before determining data types, preventing incorrect parsing and type warnings.


In [10]:
# Relative path handling: works whether launched from the workspace root or inside the notebooks folder
data_path = 'data/raw/Train_Dataset.csv' if os.path.exists('data/raw/Train_Dataset.csv') else '../data/raw/Train_Dataset.csv'

# Load raw dataset
df = pd.read_csv(data_path, low_memory=False)

print(f"Dataset loaded successfully from: '{data_path}'")


Dataset loaded successfully from: '../data/raw/Train_Dataset.csv'


## 3. Basic Metadata & Dataset Dimensions

Let's inspect the high-level dimensions of our dataset:
- **`df.shape`** returns a tuple `(number_of_rows, number_of_columns)`.
  - **Rows (Records/Samples):** Each row represents an individual customer application for an automobile loan.
  - **Columns (Features/Attributes):** Each column represents a property of the applicant or loan (e.g., income, loan amount, car ownership, repayment history).


In [11]:
# Check dataset shape
num_rows, num_cols = df.shape
print(f"Dataset Shape: {df.shape}")
print(f"Total Number of Rows (Loan Applicants): {num_rows:,}")
print(f"Total Number of Columns (Features):     {num_cols}")


Dataset Shape: (121856, 40)
Total Number of Rows (Loan Applicants): 121,856
Total Number of Columns (Features):     40


### 3.1 First Look at the Data (`head()`)
`df.head()` displays the first 5 rows of the DataFrame. This gives us a quick preview of what actual values look like across columns.


In [12]:
# Display first 5 rows
df.head()


,ID,Client_Income,Car_Owned,Bike_Owned,Active_Loan,House_Own,Child_Count,Credit_Amount,Loan_Annuity,Accompany_Client,Client_Income_Type,Client_Education,Client_Marital_Status,Client_Gender,Loan_Contract_Type,Client_Housing_Type,Population_Region_Relative,Age_Days,Employed_Days,Registration_Days,ID_Days,Own_House_Age,Mobile_Tag,Homephone_Tag,Workphone_Working,Client_Occupation,Client_Family_Members,Cleint_City_Rating,Application_Process_Day,Application_Process_Hour,Client_Permanent_Match_Tag,Client_Contact_Work_Tag,Type_Organization,Score_Source_1,Score_Source_2,Score_Source_3,Social_Circle_Default,Phone_Change,Credit_Bureau,Default
0,12142509,6750,0.0,0.0,1.0,0.0,0.0,61190.55,3416.85,Alone,Commercial,Secondary,M,Male,CL,Home,0.028663,13957,1062,6123,383,NaN,1,1,0,Sales,2.0,2.0,6.0,17.0,Yes,Yes,Self-employed,0.568066,0.478787,NaN,0.0186,63.0,NaN,0
1,12138936,20250,1.0,0.0,1.0,NaN,0.0,15282,1826.55,Alone,Service,Graduation,M,Male,CL,Home,0.008575,14162,4129,7833,21,0.0,1,0,1,NaN,2.0,2.0,3.0,10.0,Yes,Yes,Government,0.563360,0.215068,NaN,NaN,NaN,NaN,0
2,12181264,18000,0.0,0.0,1.0,0.0,1.0,59527.35,2788.2,Alone,Service,Graduation dropout,W,Male,CL,Family,0.0228,16790,5102,NaN,331,NaN,1,0,0,Realty agents,2.0,2.0,4.0,NaN,Yes,Yes,Self-employed,NaN,0.552795,0.329655054,0.0742,277.0,0.0,0
3,12188929,15750,0.0,0.0,1.0,1.0,0.0,53870.4,2295.45,Alone,Retired,Secondary,M,Male,CL,Home,0.010556,23195,365243,NaN,775,NaN,1,0,0,NaN,2.0,3.0,2.0,15.0,Yes,Yes,XNA,NaN,0.135182,0.631354537,NaN,1700.0,3.0,0
4,12133385,33750,1.0,0.0,1.0,0.0,2.0,133988.4,3547.35,Alone,Commercial,Secondary,M,Female,CL,Home,0.020713,11366,2977,5516,4043,6.0,1,0,0,Laborers,4.0,1.0,3.0,NaN,Yes,Yes,Business Entity Type 3,0.508199,0.301182,0.355638717,0.2021,674.0,1.0,0


### 3.2 Dataset Information (`df.info()`)
`df.info()` provides a concise summary of the DataFrame:
- **Column names**
- **Non-null count:** How many entries in that column are not missing.
- **Data type (Dtype):**
  - `int64`: Whole numbers (integers).
  - `float64`: Decimal numbers.
  - `object` / `string`: Text or mixed data types.
- **Memory usage:** How much RAM the dataset occupies.


In [13]:
# Display detailed column metadata and memory usage
df.info(verbose=True, show_counts=True)


<class 'pandas.DataFrame'>
RangeIndex: 121856 entries, 0 to 121855
Data columns (total 40 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   ID                          121856 non-null  int64  
 1   Client_Income               118249 non-null  str    
 2   Car_Owned                   118275 non-null  float64
 3   Bike_Owned                  118232 non-null  float64
 4   Active_Loan                 118221 non-null  float64
 5   House_Own                   118195 non-null  float64
 6   Child_Count                 118218 non-null  float64
 7   Credit_Amount               118224 non-null  str    
 8   Loan_Annuity                117044 non-null  str    
 9   Accompany_Client            120110 non-null  str    
 10  Client_Income_Type          118155 non-null  str    
 11  Client_Education            118211 non-null  str    
 12  Client_Marital_Status       118383 non-null  str    
 13  Client_Gender            

## 4. Missing Values Analysis
In real-world data collection, applicants may skip optional questions, records might be lost, or systems might fail to capture certain fields. In Pandas, missing data is represented as `NaN` (Not a Number) or `None`.

Here, we:
1. Count the number of missing values per column with `df.isnull().sum()`.
2. Compute the percentage of missing values: `(missing_count / total_rows) * 100`.
3. Display all columns containing missing values, sorted from highest to lowest percentage.


In [14]:
# Count missing values per column
missing_count = df.isnull().sum()

# Calculate percentage of missing values
missing_pct = (missing_count / len(df)) * 100

# Combine into a clean summary DataFrame
missing_summary = pd.DataFrame({
    'Column_Name': df.columns,
    'Missing_Count': missing_count.values,
    'Missing_Percentage (%)': missing_pct.values.round(2)
})

# Filter only columns that have at least one missing value and sort descending
missing_summary_filtered = (
    missing_summary[missing_summary['Missing_Count'] > 0]
    .sort_values(by='Missing_Percentage (%)', ascending=False)
    .reset_index(drop=True)
)

print(f"Total columns with missing values: {len(missing_summary_filtered)} out of {num_cols}")
missing_summary_filtered


Total columns with missing values: 33 out of 40


,Column_Name,Missing_Count,Missing_Percentage (%)
0,Own_House_Age,80095,65.73
1,Score_Source_1,68835,56.49
2,Social_Circle_Default,61928,50.82
3,Client_Occupation,41435,34.00
4,Score_Source_3,26921,22.09
5,Credit_Bureau,18540,15.21
6,ID_Days,5968,4.90
7,Score_Source_2,5686,4.67
8,Population_Region_Relative,4857,3.99
9,Loan_Annuity,4812,3.95


### 💡 Plain English Interpretation of Missing Values:
1. **High Missingness (>50%):**
   - `Own_House_Age` (~66% missing): This column has high missingness primarily because not all applicants own a house (indicated by `House_Own`). If someone rents or lives with parents, their home age is naturally absent.
   - `Score_Source_1` (~56% missing): An external credit bureau score. Many applicants simply do not have an available score from this specific agency.
2. **Moderate Missingness (10% - 50%):**
   - Features like `Social_Circle_Default`, `Client_Occupation`, and `Score_Source_3` have missing rates between 18% and 53%.
3. **Low Missingness (<5%):**
   - Features like `Credit_Bureau`, `Phone_Change`, `Loan_Annuity`, and `Credit_Amount` have very few missing records (<5%).
4. **Action for Preprocessing:**
   - In `03_data_preprocessing.ipynb`, we will apply appropriate strategies: imputation (mean, median, mode) for moderate/low missingness, indicator flags, or domain-specific handling.


## 5. Identifying Numerical Columns Stored Incorrectly as Object / String

In tabular data mining, numerical features (such as income, loan amount, age, and scores) must be stored as numbers (`int64` or `float64`) so that:
- We can perform mathematical operations (mean, median, standard deviation).
- Machine learning algorithms can compute distances and weights.

However, if even a **single row** in a column contains a non-numeric character (such as `'$'`, `'#'`, `'@'`, `'x'`, or `'#VALUE!'`), Pandas will parse the **entire column** as `object` or `string`!

Let's find out which columns are affected and examine what dirty characters are causing this issue.


In [16]:
# Expected numerical columns that should be int or float based on domain meaning
candidate_numerical_cols = [
    'Client_Income',
    'Credit_Amount',
    'Loan_Annuity',
    'Population_Region_Relative',
    'Age_Days',
    'Employed_Days',
    'Registration_Days',
    'ID_Days',
    'Score_Source_3'
]

# Check their current data types in the loaded DataFrame
dirty_columns_analysis = []

for col in candidate_numerical_cols:
    current_dtype = str(df[col].dtype)
    
    # Try converting to numeric using pd.to_numeric with errors='coerce'
    # Any value that cannot be converted to a number becomes NaN
    coerced_series = pd.to_numeric(df[col], errors='coerce')
    
    # Identify entries where the original was NOT null, but conversion resulted in NaN
    corrupted_mask = df[col].notna() & coerced_series.isna()
    corrupted_count = corrupted_mask.sum()
    unique_corrupted_values = df.loc[corrupted_mask, col].unique().tolist()
    
    dirty_columns_analysis.append({
        'Feature': col,
        'Current_Dtype': current_dtype,
        'Expected_Dtype': 'float64 / int64',
        'Corrupted_Rows_Count': corrupted_count,
        'Dirty_Characters_Found': unique_corrupted_values
    })

dirty_cols_df = pd.DataFrame(dirty_columns_analysis)
dirty_cols_df


,Feature,Current_Dtype,Expected_Dtype,Corrupted_Rows_Count,Dirty_Characters_Found
0,Client_Income,str,float64 / int64,15,[$]
1,Credit_Amount,str,float64 / int64,5,[$]
2,Loan_Annuity,str,float64 / int64,14,"[$, #VALUE!]"
3,Population_Region_Relative,str,float64 / int64,11,"[@, #]"
4,Age_Days,str,float64 / int64,17,[x]
5,Employed_Days,str,float64 / int64,17,[x]
6,Registration_Days,str,float64 / int64,17,[x]
7,ID_Days,str,float64 / int64,17,[x]
8,Score_Source_3,str,float64 / int64,1,[&]


### 🔎 Inspecting Sample Corrupted Rows
Let's see some concrete examples of rows containing these invalid characters to understand what happened during data entry or CSV export.


In [17]:
from IPython.display import display

# Show sample rows where Client_Income has non-numeric characters
print("--- Sample rows with corrupted 'Client_Income' ---")
display(df[df['Client_Income'] == '$'][['ID', 'Client_Income', 'Client_Income_Type', 'Default']].head(3))

# Show sample rows where Age_Days or Employed_Days has non-numeric characters
print("\n--- Sample rows with corrupted 'Age_Days' or 'Employed_Days' ---")
display(df[(df['Age_Days'] == 'x') | (df['Employed_Days'] == 'x')][['ID', 'Age_Days', 'Employed_Days', 'Default']].head(3))

# Show sample rows where Score_Source_3 has non-numeric characters
print("\n--- Sample rows with corrupted 'Score_Source_3' ---")
display(df[df['Score_Source_3'] == '&'][['ID', 'Score_Source_3', 'Default']].head(3))


--- Sample rows with corrupted 'Client_Income' ---


,ID,Client_Income,Client_Income_Type,Default
1586,12212892,$,Service,0
9963,12116280,$,Service,0
22639,12132177,$,Commercial,0



--- Sample rows with corrupted 'Age_Days' or 'Employed_Days' ---


,ID,Age_Days,Employed_Days,Default
16335,12182976,x,x,0
30043,12193379,x,x,0
35293,12188608,x,x,0



--- Sample rows with corrupted 'Score_Source_3' ---


,ID,Score_Source_3,Default
28193,12105215,&,1


### 💡 Plain English Explanation for Undergraduates:
Why did these columns get classified as `object`?
- **`Client_Income` & `Credit_Amount`:** Contain `$` characters instead of pure numbers. When someone entered `$` or exported from Excel with raw currency formatting, Pandas couldn't treat it as a number.
- **`Loan_Annuity`:** Contains `$` symbols and `'#VALUE!'` (a classic formula error from Microsoft Excel).
- **`Age_Days`, `Employed_Days`, `Registration_Days`, `ID_Days`:** Contain `'x'` characters, likely used by data entry operators as a placeholder for "unknown" or "unverified".
- **`Population_Region_Relative`:** Contains `'@'` and `'#'` special characters.
- **`Score_Source_3`:** Contains `'&'` characters.

**Key Learning Point:**  
Before feeding data into machine learning models, these invalid characters must be cleaned or replaced with `NaN`, and the columns must be converted to float/integer (`pd.to_numeric()`). This will be performed in our Data Preprocessing stage (`03_data_preprocessing.ipynb`).


## 6. Target Variable Inspection: `Default`
In this project, our goal is to predict whether a customer will default on their vehicle loan.
- **`Default = 1`:** The client defaulted on payments (high risk).
- **`Default = 0`:** The client repaid on time (low risk).

Let's check the distribution of the target class.


In [19]:
# Target variable class distribution
default_counts = df['Default'].value_counts()
default_percentages = df['Default'].value_counts(normalize=True) * 100

target_df = pd.DataFrame({
    'Count': default_counts,
    'Percentage (%)': default_percentages.round(2)
})
target_df.index = ['Non-Default (0)', 'Default (1)']
target_df


,Count,Percentage (%)
Non-Default (0),112011,91.92
Default (1),9845,8.08


### 💡 Target Variable Takeaway:
- Over 91% of applicants did not default (`0`), while approximately 8% defaulted (`1`).
- This means our dataset is **imbalanced**. Class imbalance is typical in credit and fraud detection, and we will address it using appropriate evaluation metrics (ROC-AUC, Precision, Recall, F1-Score) and modeling techniques in upcoming notebooks.


## 7. Summary & Next Steps

### Summary of What We Learned:
1. **Dataset Scale:** 121,856 rows and 40 columns.
2. **Missing Data:** 19 columns contain missing values. `Own_House_Age` and `Score_Source_1` have the highest percentage of missing values.
3. **Data Type Anomalies:** Several key numeric columns (`Client_Income`, `Credit_Amount`, `Loan_Annuity`, `Age_Days`, `Employed_Days`, `Score_Source_3`, etc.) are improperly parsed as strings because of corrupted characters (`$`, `x`, `&`, `#VALUE!`, `@`, `#`).
4. **Target Distribution:** Imbalanced binary classification problem (~92% non-default vs. ~8% default).

### Next Notebook:
👉 In **`02_eda.ipynb`**, we will perform Exploratory Data Analysis, visualize distributions, examine correlations, and study customer behaviors associated with loan default.
